lib/Platform/Intl/java/com/facebook/hermes/intl/Intl.java

```diff
@@ -8,45 +8,306 @@
+        // Even though they have similar APIs, there are subtle deviations for e.g. in error handling.
+        // Unlike the icu4c equivalents, the forLanguageTag in icu4j doesn't report any errors while parsing ..
+        // For e.g. icu4c identifies bogus locids.. and report failure when part of the loclid can't be parsed. (https://github.com/unicode-org/icu/blob/79fac501010d63231c258dc0d4fb9a9e87ddb8d8/icu4c/source/common/locid.cpp#L816)
+        // This results in our implementation a bit more leniant (i.e. not throwing RangeError for certain invalid inputs) campared to V8/Chakra
+        // Both icu4c and icu4j implementations use private tables and mappings embedded in code for special cases such as deprecated and grandfathered  language code.
+        // icu4c: https://github.com/unicode-org/icu/blob/9219c6ae038a3556dffca880e93d2f2ca00e685a/icu4c/source/common/uloc_tag.cpp#L103
+        // icu4j: https://github.com/unicode-org/icu/blob/79fac501010d63231c258dc0d4fb9a9e87ddb8d8/icu4j/main/classes/core/src/com/ibm/icu/impl/locale/LanguageTag.java#L43
+        // Android-icu4j: https://android.googlesource.com/platform/external/icu/+/refs/heads/master/android_icu4j/
+        //
+        // Clearly icu4c implementation has a few more private tables implementated, for e.g. the deprecated languages and regions. This results in subtle behavioural differences between our implemenation and say, V8.
+        //
+        // As of now, Firefox seems to be by far the most compliant engine out there.
+        //
+        // Our current implementation is directly over icu4j available with Android platform, as hasn't implemented any private tables/mappings yet.
+        // We fail many tests in our testcases based on https://github.com/tc39/test262/tree/master/test/intl402/Intl, but mostly special scenarios.
+        // Our implementation closely follows the current V8 implementation, and our differences from V8 stems from
+        // 1. The behavioural difference between icu4j and icu4c
+        // 2. We didnt' implement some of the private tables and mappings other than what we get from ICU. (for e.g. https://github.com/v8/v8/blob/d432b2185ca9f2c1c405df872e5546cd2cf2dfd0/src/objects/intl-objects.cc#L802)
+        // The failed test cases are commented with TODO::Hermes tag in intl_getCanonicalNames_*.js
+        //
+        // Ref: https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/ULocale.html#forLanguageTag-java.lang.String-
+        // Ref: https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1Locale.html#af76028775e37fd75a30209aaede551e2
+
+        ULocale uLocale= ULocale.forLanguageTag(locale);
+        String tag = uLocale.toLanguageTag();
+        return tag;
+
+        // Note: TODO : An alternative implementation could use https://developer.android.com/reference/android/icu/util/ULocale.Builder to build the locale.. which does structural validation internally based on documentation.
```

Comment:
This seems like it would reduce the java code size, by delegating the validation to the library.  Is there a reason to prefer doing it the way it is here?

Reply:
In the latest iteration, we've switcher to Builder, but it comes with its own deficiencies. For e.g. the Builder implementation in icu4j don't look up the grandfathered tag mapping. But, it has the performance advantage as it doesn't require parsing the whole tag again, which we have already parsed.

This is a super critical aspect about this change. Currently, we are adding a set of autogenerated data into the binary for improving the correctness, but resulting in bigger binary size. Having predictable behaviour is important. I completely understood your concern. I think, we should review the details about the implementation in the coming weeks. For now, i'd like to go ahead with the current implementation.

Reply:
That sounds fine.  It will be easier to consider size/functionality tradeoffs when the system is more mature.